swarm/swap_test.go

```diff
@@ -0,0 +1,468 @@
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+var (
+	printStats     = flag.Bool("printstats", false, "print swap stats")
+	bucketKeySwarm = simulation.BucketKey("swarm")
+)
+
+//In TestSwapNetworkSymmetricFileUpload we set up a network with arbitrary number of nodes
+//(16), and each of the nodes uploads a file of same size
+//Afterwards we check that every node's balance WITH ANOTHER PEER
+//has the same value but opposite sign
+func TestSwapNetworkSymmetricFileUpload(t *testing.T) {
+	//default hardcoded network size
+	nodeCount := 16
+	//every node has a map to all nodes it had interactions
+	//each entry in the map is a map of the other node with all the balances
+	balancesMap := make(map[enode.ID]map[enode.ID]int64)
+
+	//setup the simulation
+	//use a complete node setup via `NewSwam`
+	sim := simulation.New(map[string]simulation.ServiceFunc{
+		"swarm": func(ctx *adapters.ServiceContext, bucket *sync.Map) (s node.Service, cleanup func(), err error) {
+			config := api.NewConfig()
+			config.Port = ""
+
+			dir, err := ioutil.TempDir("", "swap-network-test-node")
+			if err != nil {
+				return nil, nil, err
+			}
+			cleanup = func() {
+				err := os.RemoveAll(dir)
+				if err != nil {
+					log.Error("cleaning up swarm temp dir", "err", err)
+				}
+			}
+
+			config.Path = dir
+
+			privkey, err := crypto.GenerateKey()
+			if err != nil {
+				return nil, cleanup, err
+			}
+
+			config.Init(privkey)
+
+			//set Swap to be enabled for this test
+			config.SwapEnabled = true
+
+			swarm, err := NewSwarm(config, nil)
+			if err != nil {
+				return nil, cleanup, err
+			}
+
+			bucket.Store(bucketKeySwarm, swarm)
+			log.Info("new swarm", "bzzKey", config.BzzKey, "baseAddr", fmt.Sprintf("%x", swarm.bzz.BaseAddr()))
+			return swarm, cleanup, nil
+		},
+	})
+
+	ctx := context.Background()
+	files := make([]file, 0)
+
+	//upload a snapshot
+	err := sim.UploadSnapshot(fmt.Sprintf("network/stream/testing/snapshot_%d.json", nodeCount))
+	if err != nil {
+		t.Fatal(err)
+	}
+	//run the simulation
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) error {
+		//wait for kademlia to be healthy
+		if _, err := sim.WaitTillHealthy(ctx, 2); err != nil {
+			return err
+		}
+
+		nodeIDs := sim.UpNodeIDs()
+		rand.Shuffle(len(nodeIDs), func(i, j int) {
+			nodeIDs[i], nodeIDs[j] = nodeIDs[j], nodeIDs[i]
+		})
+		//upload a file for every node
+		for _, id := range nodeIDs {
+			item, ok := sim.NodeItem(id, bucketKeySwarm)
+			if !ok {
+				log.Error("No swarm")
+				return errors.New("No swarm")
+			}
+			swarm := item.(*Swarm)
+			key, data, err := uploadFile(swarm)
+			if err != nil {
+				return err
+			}
+			log.Trace("file uploaded", "node", id, "key", key.String())
+			files = append(files, file{
+				addr:   key,
+				data:   data,
+				nodeID: id,
+			})
+		}
+
+		// File retrieval check is repeated until all uploaded files are retrieved from all nodes
+		// or until the timeout is reached.
+		for {
```

Comment:
```
for retrieveForSwap(sim, files) != 0 {}
```

Reply:
for some obscure reason if I use this notation the tests sometimes fail....